src/openrct2/world/footpath.c

```diff
@@ -745,7 +745,22 @@ money32 footpath_provisional_set(sint32 type, sint32 x, sint32 y, sint32 z, sint
 
     if (!scenery_tool_is_active())
     {
-        map_set_virtual_floor_height(z * 8);
+        if (cost == MONEY32_UNDEFINED)
+        {
+            // If we can't build this, don't show a virtual floor.
+            map_set_virtual_floor_height(0);
+        }
+        else if (gFootpathConstructSlope == TILE_ELEMENT_SLOPE_FLAT
+             || gFootpathProvisionalPosition.z * 8 < gFootpathConstructFromPosition.z)
+        {
+            // Going either straight on, or down.
+            map_set_virtual_floor_height(gFootpathProvisionalPosition.z * 8);
+        }
+        else
+        {
+            // Going up in the world!
+            map_set_virtual_floor_height((gFootpathProvisionalPosition.z + 1) * 8 + 4);
```

Comment:
If we want the virtual floor to be at the top of the slope (like it's at the bottom when going down), this should be `map_set_virtual_floor_height((gFootpathProvisionalPosition.z + 2) * 8);`